provider/schema_light.go

```diff
@@ -0,0 +1,238 @@
+		return false
+	}
+	if !ls.isDanglingRef(ref) {
+		return false
+	}
+	return true
+}
+
+// Recognize { "type": "string", "$ref": "#/types/aws:rds/engineType:EngineType"}
+func (ls *lightSchema) isDanglingStringRef(m map[string]any) bool {
+	if !ls.isTypeString(m) {
+		return false
+	}
+	if !ls.isDanglingPlainRef(m) {
+		return false
+	}
+	return true
+}
+
+// Recognize unions of type String+R where R is dangling, such as:
+//
+//		{
+//	             "type": "string",
+//			"oneOf": [
+//			    {
+//				"type": "string"
+//			    },
+//			    {
+//				"type": "string",
+//				"$ref": "#/types/aws:iam/role:Role"
+//			    }
+//			]
+//		}
+func (ls *lightSchema) isUnionStringDanglingRef(m map[string]any) bool {
+	if !ls.isTypeString(m) {
+		return false
+	}
+	oneOf, ok := m["oneOf"]
+	if !ok {
+		return false
+	}
+	cases, ok := oneOf.([]any)
+	if !ok {
+		return false
+	}
+	if len(cases) != 2 {
+		return false
+	}
+	if !reflect.DeepEqual(cases[0], map[string]any{"type": "string"}) {
+		return false
+	}
+	case1, ok := cases[1].(map[string]any)
+	if !ok {
+		return false
+	}
+	if !ls.isDanglingStringRef(case1) {
+		return false
+	}
+	return true
+}
+
+// Recognize unions of two dangling refs R1+R2 where R is stringy:
+//
+//	{
+//		"oneOf": [
+//		    {
+//		    	"$ref": "#/types/aws:iam/Role:Role"
+//		    },
+//		    {
+//			"type": "string",
+//			"$ref": "#/types/aws:iam/role:Role"
+//		    }
+//		]
+//	}
+func (ls *lightSchema) isUnion2DanglingRefs(m map[string]any) bool {
+	oneOf, ok := m["oneOf"]
+	if !ok {
+		return false
+	}
+	cases, ok := oneOf.([]any)
+	if !ok {
+		return false
+	}
+	if len(cases) != 2 {
+		return false
+	}
+	case0, ok := cases[0].(map[string]any)
+	if !ok {
+		return false
+	}
+	case1, ok := cases[1].(map[string]any)
+	if !ok {
+		return false
+	}
+	if !ls.isDanglingPlainRef(case0) {
+		return false
+	}
+	if !ls.isDanglingStringRef(case1) {
+		return false
+	}
+	return true
+}
+
+func (ls *lightSchema) rewrite(schema any) any {
+	switch x := schema.(type) {
+	case map[string]any:
+		switch {
+		// { "type": "string", "oneOf": [String, DanglingRef] } --> { "type": "string" }
+		case ls.isUnionStringDanglingRef(x):
+			delete(x, "oneOf")
+			return x
+		// { "oneOf": [DanglingRef1, StringDanglingRef2] } --> { "type": "string" }
+		case ls.isUnion2DanglingRefs(x):
+			if !ls.isTypeString(x) {
+				_, gotType := x["type"]
+				contract.Assertf(!gotType, "refusing to override type")
+				x["type"] = "string"
+			}
+			delete(x, "oneOf")
+			return x
+		// { "type": "string", "$ref": "#/types/aws:rds/engineType:EngineType"} --> { "type": "string" }
+		case ls.isDanglingStringRef(x):
+			delete(x, "$ref")
+			return x
+		default:
+			for k, v := range x {
+				x[k] = ls.rewrite(v)
+			}
+			return x
+		}
+
+	case []any:
+		for i, e := range x {
+			x[i] = ls.rewrite(e)
+		}
+		return x
+	default:
+		return schema
+	}
+}
+
+func (ls *lightSchema) write(lightSchemaFile string) {
+	bytes, err := json.Marshal(ls.PackageSpec)
```

Comment:
Should we `MarshalIndent` so we can reason about diffs in future PRs?

Reply:
Interestingly we don't do that for the official schema as it's fairly large. I'd just maybe exclude it from GitHub diffs so it does not show in the UI while still being tracked in history?